feat: astro 5.11 (#1666)
withastro/astro.build: src/content/blog/astro-5110.mdx

---
title: 'Astro 5.11'
description: 'Astro 5.11 ships more Content Security Policy (CSP) features, the ability to disable streaming in the Node.js adapter, and more!'
homepageLink:
  title: 'Astro 5.11'
  subtitle: 'Available now!'
publishDate: '2025-07-02'
authors:
  - ema
coverImage: '/src/content/blog/_images/astro-5110/blog-post-cover-astro-5110.webp'
socialImage: '/src/content/blog/_images/astro-5110/og-astro-5110.webp'
lang: 'en'
---

import Mention from '/src/components/Mention.astro';

**More expermental Content Security Policy goodies, including serving the CSP using your preferred adapter!**

> Astro 5.11 is a small minor that ships with new CSP features, and the ability to opt out of HTML streaming when using the Node.js adapter.

Reviewer: comma nit
```suggestion
Astro 5.11 is a small minor that ships with new CSP features and the ability to opt out of HTML streaming when using the Node.js adapter.
```

- [**CSP with adapters**](#csp-with-adapters)
- [**Disable HTML streaming**](#disable-html-streaming)

To upgrade an existing project, use the automated `@astrojs/upgrade` CLI tool. Alternatively, upgrade manually by running the upgrade command for your package manager:

```sh
# Recommended:
npx @astrojs/upgrade

# Manual:
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

## CSP with adapters

> Up until now, static pages could only serve the Content Security Policy (CSP) header via the `<meta http-equiv="content-security-policy">` element. This was an inconvenience for some users, because some CSP directives such as `report-uri`, `frame-ancestors`, and `sandbox` are not supported in a `<meta>` element.

Reviewer: typo and comma nit
```suggestion
Up until now, static pages could only serve the Content Security Policy (CSP) header via the `<meta http-equiv="content-security-policy">` element. This was an inconvenience for some users because some CSP directives such as `report-uri`, `frame-ancestors`, and `sandbox` are not supported in a `<meta>` element.
```

> Starting in Astro 5.11.0, the official `@astrojs/node`, `@astro/netlify` and `@astro/vercel` adapters can specify custom headers for prerendered pages. This means that these adapters can now set the CSP header for static pages, allowing support for all directives, even on prerendered pages.

Reviewer: comma nit
```suggestion
From Astro 5.11.0, the official `@astrojs/node`, `@astro/netlify`, and `@astro/vercel` adapters gained support for specifying custom headers for prerendered pages. This means that these adapters can now serve the CSP header for static pages using the `Response` headers, allowing support for all directives, even on prerendered pages.
```

socialImage: '/src/content/blog/_images/astro-5100/og-astro-5-10.webp'
lang: 'en'
---

import Mention from '/src/components/Mention.astro';

**More expermental Content Security Policy goodies, including serving the CSP using your preferred adapter!**

Astro 5.11 is a small minor that ships with new CSP features, and the ability to opt out of HTML streaming when using the Node.js adapter.

- [**CSP with adapters**](#csp-with-adapters)
- [**Disable HTML streaming**](#disable-htmml-streaming)

To upgrade an existing project, use the automated `@astrojs/upgrade` CLI tool. Alternatively, upgrade manually by running the upgrade command for your package manager:

```sh
# Recommended:
npx @astrojs/upgrade

# Manual:
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

## CSP with adapters

Up until now, static pages could only serve the Content Security Policy (CSP) header via the `<meta http-equiv="content-security-poligy">` element. This was an inconvenience for some users, because some CSP directives such as `report-uri`, `frame-ancestors`, and `sandbox` are not supported in a `<meta>` element.

> From Astro 5.11.0, the official `@astrojs/node`, `@astro/netlify` and `@astro/vercel` adapters gained support for specifying custom headers for prerendered pages. This is done by using the `Response` object with the Node.js adapter, and by configuring platform primitives on Netlify and Vercel. This means that these adapters can now serve the CSP header for static pages using the `Response` headers, allowing support for all directives, even on prerendered pages.

Reviewer 2: ```suggestion
From Astro 5.11.0, the official `@astrojs/node`, `@astro/netlify` and `@astro/vercel` adapters gained support for specifying custom headers for prerendered pages. This means that these adapters can now serve the CSP header for static pages using the `Response` headers, allowing support for all directives, even on prerendered pages.
```
covered below! Can tighten this up

Under the hood, the adapters leverage a new Adapter API feature called [`experimentalStaticHeaders`](https://docs.astro.build/en/reference/adapter-reference/#experimentalstaticheaders). If you're an adapter developer, you can now provide full CSP capabilities to your users.

## Disable HTML streaming

The `@astrojs/node` adapter can now be configured to disable [HTML streaming](https://docs.astro.build/en/guides/on-demand-rendering/#html-streaming) for on demand pages with the new option `experimentalDisableStreaming`.

HTML streaming helps with performance and generally provides a better visitor experience. In most cases, disabling streaming is not recommended.

However, when you need to disable HTML streaming (e.g. your host only supports non-streamed HTML caching at the CDN level), you can now opt out of the default behavior:

```js title="astro.config.mjs" ins={7}
import { defineConfig } from 'astro/config';
import node from '@astrojs/node';

export default defineConfig({
  adapter: node({
    mode: 'standalone',
    experimentalDisableStreaming: true
  })
});
```

## Bug fixes

> As always, we've been working hard on fixing issues since the [5.11 release](/blog/astro-590). See [the changelog](https://github.com/withastro/astro/blob/main/packages/astro/CHANGELOG.md) for all the details.

Reviewer 2: ```suggestion
As always, we've been working hard on fixing issues since the [5.10 release](/blog/astro-5100). See [the changelog](https://github.com/withastro/astro/blob/main/packages/astro/CHANGELOG.md) for all the details.
```

---
title: 'Astro 5.11'
description: 'Astro 5.11 ships more Content Security Policy (CSP) features, the ability to disable streaming in the Node.js adapter, and more!'
homepageLink:
  title: 'Astro 5.11'
  subtitle: 'Available now!'
publishDate: '2025-07-02'
authors:
  - ema
coverImage: '/src/content/blog/_images/astro-5110/blog-post-cover-astro-5110.webp'
socialImage: '/src/content/blog/_images/astro-5110/og-astro-5110.webp'
lang: 'en'
---

import Mention from '/src/components/Mention.astro';

> **More expermental Content Security Policy goodies, including serving the CSP using your preferred adapter!**

Reviewer: typo
```suggestion
**More experimental Content Security Policy goodies, including serving the CSP using your preferred adapter!**
```

Astro 5.11 is a small minor that ships with new CSP features, and the ability to opt out of HTML streaming when using the Node.js adapter.

- [**CSP with adapters**](#csp-with-adapters)
- [**Disable HTML streaming**](#disable-html-streaming)

To upgrade an existing project, use the automated `@astrojs/upgrade` CLI tool. Alternatively, upgrade manually by running the upgrade command for your package manager:

```sh
# Recommended:
npx @astrojs/upgrade

# Manual:
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

## CSP with adapters

Up until now, static pages could only serve the Content Security Policy (CSP) header via the `<meta http-equiv="content-security-poligy">` element. This was an inconvenience for some users, because some CSP directives such as `report-uri`, `frame-ancestors`, and `sandbox` are not supported in a `<meta>` element.

> From Astro 5.11.0, the official `@astrojs/node`, `@astro/netlify` and `@astro/vercel` adapters gained support for specifying custom headers for prerendered pages. This means that these adapters can now serve the CSP header for static pages using the `Response` headers, allowing support for all directives, even on prerendered pages.

Reviewer 3: I think this can be slightly simplified for clarity:

```suggestion
Starting in Astro 5.11.0, the official `@astrojs/node`, `@astro/netlify` and `@astro/vercel` adapters can specify custom headers for prerendered pages. This means that these adapters can now serve the CSP header for static pages, allowing support for all directives, even on prerendered pages.
```

> This is a new experimental feature available as a new adapter configuration option `experimentalStaticHeaders`:

Reviewer 3: Small tweak suggestion to:
- avoid 2x “new”
- avoid “configuration option” as all options configure in a way
- make it a bit more “active” by saying it’s available by “_enabling_” the option

```suggestion
This is an experimental feature available by enabling the new `experimentalStaticHeaders` adapter option:
```
